cli/silence_import.go

```diff
@@ -0,0 +1,67 @@
+package cli
+
+import (
+	"os"
+	"io/ioutil"
+	"encoding/json"
+	"fmt"
+
+	"github.com/prometheus/alertmanager/types"
+	"github.com/spf13/cobra"
+	"github.com/pkg/errors"
+)
+
+var importCmd = &cobra.Command{
+	Use:   "import [JSON file]",
+	Short: "Import silences",
+	Long: `Import alertmanager silences from JSON file or stdin
+
+  This command can be used to bulk import silences from a JSON file
+  created by query command. For example:
+
+  amtool silence query -o json foo > foo.json
+  amtool silence import foo.json
+
+  JSON data can also come from stdin if no param is specified.
+	`,
+	Args: cobra.MaximumNArgs(1),
+	Run:  CommandWrapper(bulkImport),
+}
+
+func bulkImport(cmd *cobra.Command, args []string) error {
+	input := os.Stdin
+	var err error
+
+	if len(args) == 1 {
+		input, err = os.Open(args[0])
+
+		if err != nil {
+			return err
+		}
+	}
+
+	data, err := ioutil.ReadAll(input)
```

Comment:
`ioutil.ReadAll` will load the entire input into memory. It would be better to process the information as a stream.

See https://golang.org/pkg/encoding/json/#example_Decoder_Decode_stream

Reply:
Actually it ended up being a super easy change:

```go
dec := json.NewDecoder(input)

t, err := dec.Token()
if err != nil {
	return err
}

for dec.More() {
	var s types.Silence
	err := dec.Decode(&s)
	if err != nil {
		log.Fatal(err)
	}

	s.ID = ""

	err = addSilence(s)
	if err != nil {
		msg := fmt.Sprintf("couldn't add silence: %s", s)
		return errors.Wrap(err, msg)
	}
}

t, err = dec.Token()
if err != nil {
	return err
}

return nil
```